fiksi/src/elements/mod.rs

```diff
@@ -29,6 +39,39 @@ pub(crate) mod element {
         pub(crate) fn drop_system_id(self) -> ElementId {
             ElementId { id: self.id }
         }
+
+        /// Get a type-erased handle to the element.
+        ///
+        /// To turn the returned handle back into a typed handle, use
+        /// [`ElementHandle<AnyElement>::as_tagged_element`](ElementHandle<super::AnyElement>::as_tagged_element).
+        pub fn as_any_element(self) -> ElementHandle<super::AnyElement> {
+            ElementHandle::from_ids(self.system_id, self.id)
+        }
+    }
+
+    impl ElementHandle<super::AnyElement> {
+        /// Get a typed handle to the element.
+        pub fn as_tagged_element(self, system: &crate::System) -> TaggedElementHandle {
+            use crate::Vertex;
+
+            // TODO: return `Result` instead of panicking?
+            assert_eq!(
+                system.id, self.system_id,
+                "Tried to use an element handle that is not part of this `System`"
+            );
+
+            match system.element_vertices[self.id as usize] {
+                Vertex::Point { .. } => {
+                    TaggedElementHandle::Point(ElementHandle::from_ids(self.system_id, self.id))
+                }
+                Vertex::Line { .. } => {
+                    TaggedElementHandle::Line(ElementHandle::from_ids(self.system_id, self.id))
+                }
+                Vertex::Circle { .. } => {
+                    TaggedElementHandle::Circle(ElementHandle::from_ids(self.system_id, self.id))
+                }
+            }
+        }
```

Comment:
Perhaps `ElementHandle` should store the tag, so this method doesn't need a reference to the `System` passed in. The `PhantomData` field could become `<ElementInner::Data>` which is just `()` in all cases except for `AnyElement`.

Reply:
I've implemented this in the "Store type tag in `ElementHandle<AnyElement>`" commit. It could be a bit cleaner by using a separate struct like `AnyElementHandle`, allowing to drop the associated type that's `()` for all except `AnyElement`. In that case, to keep the APIs symmetric for both the `ElementHandle<T>` and `AnyElementHandle` case, it may be best to move methods like `System::get_element_value(handle: ElementHandle<T>)` to be `ElementHandle<T>::get_value(s: &System)` and `AnyElementHandle::get_value(s: &System)`.